flow/display_list_utils.cc

```diff
@@ -137,101 +137,210 @@ void SkMatrixDispatchHelper::reset() {
 }
 
 void ClipBoundsDispatchHelper::clipRect(const SkRect& rect,
-                                        bool isAA,
+                                        bool is_aa,
                                         SkClipOp clip_op) {
   if (clip_op == SkClipOp::kIntersect) {
-    intersect(rect);
+    intersect(rect, is_aa);
   }
 }
 void ClipBoundsDispatchHelper::clipRRect(const SkRRect& rrect,
-                                         bool isAA,
+                                         bool is_aa,
                                          SkClipOp clip_op) {
   if (clip_op == SkClipOp::kIntersect) {
-    intersect(rrect.getBounds());
+    intersect(rrect.getBounds(), is_aa);
   }
 }
 void ClipBoundsDispatchHelper::clipPath(const SkPath& path,
-                                        bool isAA,
+                                        bool is_aa,
                                         SkClipOp clip_op) {
   if (clip_op == SkClipOp::kIntersect) {
-    intersect(path.getBounds());
+    intersect(path.getBounds(), is_aa);
   }
 }
-void ClipBoundsDispatchHelper::intersect(const SkRect& rect) {
+void ClipBoundsDispatchHelper::intersect(const SkRect& rect, bool is_aa) {
   SkRect devClipBounds = matrix().mapRect(rect);
-  if (!bounds_.intersect(devClipBounds)) {
-    bounds_.setEmpty();
+  if (is_aa) {
+    devClipBounds.roundOut(&devClipBounds);
+  }
+  if (has_clip_) {
+    if (!bounds_.intersect(devClipBounds)) {
+      bounds_.setEmpty();
+    }
+  } else {
+    has_clip_ = true;
+    if (devClipBounds.isEmpty()) {
+      bounds_.setEmpty();
+    } else {
+      bounds_ = devClipBounds;
+    }
   }
 }
 void ClipBoundsDispatchHelper::save() {
-  saved_.push_back(bounds_);
+  if (!has_clip_) {
+    saved_.push_back(SkRect::MakeLTRB(0, 0, -1, -1));
+  } else if (bounds_.isEmpty()) {
+    saved_.push_back(SkRect::MakeEmpty());
+  } else {
+    saved_.push_back(bounds_);
+  }
 }
 void ClipBoundsDispatchHelper::restore() {
   bounds_ = saved_.back();
   saved_.pop_back();
+  has_clip_ = (bounds_.fLeft <= bounds_.fRight &&  //
+               bounds_.fTop <= bounds_.fBottom);
+  if (!has_clip_) {
+    bounds_.setEmpty();
+  }
+}
+void ClipBoundsDispatchHelper::reset(const SkRect* cull_rect) {
+  if ((has_clip_ = ((bool)cull_rect)) && !cull_rect->isEmpty()) {
+    bounds_ = *cull_rect;
+  } else {
+    bounds_.setEmpty();
+  }
 }
 
-void DisplayListBoundsCalculator::saveLayer(const SkRect* bounds,
-                                            bool with_paint) {
+DisplayListBoundsCalculator::DisplayListBoundsCalculator(
+    const SkRect* cull_rect)
+    : ClipBoundsDispatchHelper(cull_rect) {
+  layer_infos_.emplace_back(std::make_unique<RootLayerData>());
+  accumulator_ = layer_infos_.back()->accumulatorForLayer();
+}
+void DisplayListBoundsCalculator::setCaps(SkPaint::Cap cap) {
+  cap_is_square_ = (cap == SkPaint::kSquare_Cap);
+}
+void DisplayListBoundsCalculator::setJoins(SkPaint::Join join) {
+  join_is_miter_ = (join == SkPaint::kMiter_Join);
+}
```

Comment:
I capitalized them originally because any unclosed path has at least 2 caps and can have an arbitrary number of joins. I can change it, but the naming isn't really part of this PR.

Reply:
I added a note in the "fix up DL naming" Issue about these names.